lib/membrane/video_compositor/video_compositor.ex

```diff
@@ -0,0 +1,431 @@
+          {:ok, _} -> {:halt, true}
+          {:error, _} -> {:cont, false}
+        end
+      end)
+
+    unless started? do
+      raise "Failed to startup and connect to VideoCompositor server."
+    end
+
+    :ok
+  end
+
+  @spec system_architecture() :: :darwin_aarch64 | :darwin_x86_64 | :linux_x86_64
+  defp system_architecture() do
+    case :os.type() do
+      {:unix, :darwin} ->
+        system_architecture = :erlang.system_info(:system_architecture) |> to_string()
+
+        cond do
+          Regex.match?(~r/aarch64/, system_architecture) ->
+            :darwin_aarch64
+
+          Regex.match?(~r/x86_64/, system_architecture) ->
+            :darwin_x86_64
+
+          true ->
+            raise "Unsupported system architecture: #{system_architecture}"
+        end
+
+      {:unix, :linux} ->
+        :linux_x86_64
+
+      os_type ->
+        raise "Unsupported os type: #{os_type}"
+    end
+  end
+
+  @spec register_input_stream(input_id(), State.t(), port_number()) ::
+          port_number()
+  defp register_input_stream(input_id, state, input_port \\ 4000) do
+    if state |> State.used_ports() |> MapSet.member?(input_port) do
+      register_input_stream(input_id, state, input_port + 2)
+    else
+      case VcReq.register_input_stream(input_id, input_port, state.vc_port) do
+        :ok ->
+          input_port
+
+        {:error, %Req.Response{}} ->
+          register_input_stream(
+            input_id,
+            state,
+            input_port + 2
+          )
+
+        _other ->
+          raise "Register input failed"
+      end
+    end
+  end
+
+  @spec register_output_stream(map(), State.t(), port_number()) :: port_number()
+  defp register_output_stream(pad_options, state, output_port \\ 5000) do
+    if state |> State.used_ports() |> MapSet.member?(output_port) do
+      register_output_stream(pad_options, state, output_port + 2)
+    else
+      :ok =
+        VcReq.register_output_stream(
+          pad_options.output_id,
+          output_port,
+          pad_options.resolution,
+          pad_options.encoder_preset,
+          state.vc_port
+        )
+
+      output_port
+    end
+  end
+
+  @spec remove_input(State.t(), Membrane.Pad.ref()) :: State.t()
+  defp remove_input(state = %State{inputs: inputs}, input_ref) do
+    input_id =
+      inputs
+      |> Enum.find(fn %InputState{pad_ref: ref} -> ref == input_ref end)
+      |> then(fn %InputState{input_id: id} -> id end)
+
+    :ok = VcReq.unregister_input_stream(input_id, state.vc_port)
+
+    inputs = Enum.reject(inputs, fn %InputState{pad_ref: ref} -> ref == input_ref end)
+
+    %State{state | inputs: inputs}
+  end
+
+  @spec remove_output(State.t(), Membrane.Pad.ref()) :: State.t()
+  defp remove_output(state = %State{outputs: outputs}, output_ref) do
+    output_id =
+      outputs
+      |> Enum.find(fn %OutputState{pad_ref: ref} -> ref == output_ref end)
+      |> then(fn %OutputState{output_id: id} -> id end)
```

Comment:
Use `Map.get` instead. Do it here and in similar places

Reply:
I don't understand what you mean. 
- If you suggest using `Map.get()` to find `OutputState` with `pad_ref == ouptut_ref` I would have to first convert the list to map to just perform `get` on it (I don't want to keep this as map since I perform `find` on different fields, convert this to `context` etc.).  I think that would be less readable.
- If you suggest changing the `then` section to `Map.get()` I'm also against it. I would remove the compile time check of the `output_id` field name, making the code more error-prone.

Regardless of which line you suggested to change, I don't think it's a good idea.

Reply:
Matching on the other side is better. I will use that.